### PR TITLE
resolves #198

### LINKED ISSUE #198: Update documentation for v2
Double check docstrings for API and update hand written ``*.rst`` files in ``doc`` and ``doc/examples``.

### PR DESCRIPTION
**Pull request by lcgraham**
updates documentation and takes care of a todo

### CODE REVIEW

**Comment by smattis** [comment]
That change to `sample.py` seems to have introduced some errors.

**Comment by lcgraham** [comment]
I decided to keep the globalize parameter in ``random_sample_set`` defaulting to true because of how ``bet.sample.sample_set_base.check_num`` operates.

**Comment by lcgraham** [comment]
I think it's failing in parallel because of the changes I made to saving...

File: bet/sampling/basicSampling.py [hunk 1]
```diff
@@ -339,47 +358,35 @@ def compute_QoI_and_create_discretization(self, input_sample_set,
...
         discretization = sample.discretization(input_sample_set,
                 output_sample_set)
 
         mdat = dict()
         self.update_mdict(mdat)
 
-        if comm.rank == 0 and savefile is not None:
+        if savefile is not None:
             self.save(mdat, savefile, discretization)
         comm.barrier()
```
Comment (eecsu) [inline bet/sampling/basicSampling.py]: Won't this mean each proc saves the same file given by ``savefile``? This needs to append the ``savefile`` by the proc number, yes?

**Comment by lcgraham** [comment]
@smattis Is there any reason to continue using Shippable? Seems redundant now.

**Comment by eecsu** [comment]
I think the example file `parallel_serial.py` needs to have the ``parallel`` flag dropped from the command ``my_discretization = my_sampler.create_random_discretization(sample_type='r',
        input_obj=4, savefile="parallel_serial_example", num_samples=100,
parallel=True)``

File: bet/sample.py [hunk 2]
```diff
@@ -793,21 +844,26 @@ def save_discretization(save_disc, file_name, discretization_name=None):
...
+    if os.path.exists(local_file_name) or os.path.exists(local_file_name+'.mat'):
+        mdat = sio.loadmat(local_file_name)
+        for i, v in new_mdat.iteritems():
+            mdat[i] = v
+        if globalize and comm.rank == 0:
+            sio.savemat(local_file_name, mdat)
+        elif not globalize:
+            sio.savemat(local_file_name, mdat)
+    else:
+        if globalize and comm.rank == 0:
```
Comment (eecsu) [inline bet/sample.py]: need to define ``mdat`` in this ``else`` part.

**Comment by lcgraham** [comment]
Nevermind, Shippable and both Travis found errors but Travis thinks it passed so...

**Comment by smattis** [comment]
That is really weird that Travis says it passed when it obviously did not.

**Comment by smattis** [comment]
There's not really a reason to use both Travis and Shippable, but it gives us more comfort.

**Comment by smattis** [comment]
Hmm... Do the test pass in parallel on your machine?

**Comment by lcgraham** [comment]
The parallel tests are passing on my department cluster. I do not know why they are failing here.

**Comment by smattis** [comment]
Does one test maybe take a really long time? They are failing because nothing is being returned for 10 minutes.

**Comment by lcgraham** [comment]
The parallel tests run in about 5 min on my system. A previous run on Shippable passed everything https://app.shippable.com/runs/5776f797628e400c00283c38 . I'm going to try running the exact sequence of tests (as opposed to the parallel just alone) and see if something isn't being cleaned up from previous tests correctly.

**Comment by lcgraham** [comment]
The nosetests, mpirun -np 2 nosetests sequence worked on my machine. Does it work on your machine?

**Comment by smattis** [comment]
It appears to be freezing on a test on my ICES linux machine.

**Comment by smattis** [comment]
'''
Test :met:`bet.sampling.adaptiveSampling.sampler.generalized_chains` ... ERROR
Run :meth:`bet.sampling.adaptiveSampling.sampler.run_gen` and verify ... ERROR
ERROR
Run :meth:`bet.sampling.adaptiveSampling.sampler.run_inc_dec` and verify ... Run :meth:`bet.sampling.adaptiveSampling.sampler.run_gen` and verify ...
'''

errors and then hanging here

**Comment by smattis** [comment]
@lcgraham It is still giving errors and hanging up at the same places on my linux machine.

**Comment by lcgraham** [comment]
Yeah I that's where it's failing on Travis as well... can you send me the output of "mpirun -np 2 nosetests -x"? Is it on your UT machine? I still have access to ICES machines so I might be able to debug the ought my account there... Let me see...

**Comment by smattis** [comment]
```
mpirun -n 2 nosetests -x
......................SSSSSS......SS........................SS....SS................SS......SS....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................SSSSSSSSSSSSSS...................F
======================================================================
FAIL: Test :meth:`bet.postProcess.plotP.plot_1D_marginal_probs`.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/org/groups/chg/steven/src/BET/LG-Bet/BET/test/test_postProcess/test_plotP.py", line 189, in test_plot_marginals_1D
    nptest.assert_equal(go, True)
  File "/org/groups/chg/steven/src/anaconda2/lib/python2.7/site-packages/numpy/testing/utils.py", line 355, in assert_equal
    raise AssertionError(msg)
AssertionError: 
Items are not equal:
 ACTUAL: False
 DESIRED: True

----------------------------------------------------------------------
Ran 442 tests in 15.024s

FAILED (SKIP=15, failures=1)
```

**Comment by smattis** [comment]
Actually, I'm getting the same errors on my Macbook as well. Do you maybe have something uncommitted or in your environment that could be the culprit?

**Comment by smattis** [comment]
```
...
Tests :meth:`bet.Lp_generalized_samples.Lp_generalized_beta` ... Tests :meth:`bet.Lp_generalized_samples.Lp_generalized_beta` ... ok
ok
Test :met:`bet.sampling.adaptiveSampling.sampler.generalized_chains` ... Test :met:`bet.sampling.adaptiveSampling.sampler.generalized_chains` ... ERROR
Run :meth:`bet.sampling.adaptiveSampling.sampler.run_gen` and verify ...
```
and just hanging here

**Comment by lcgraham** [comment]
Hmmm... let me check. Also I can't access ICES machines ATM so I'll e-mail sysnet in the AM.

**Comment by lcgraham** [comment]
I had forgotten to take BET off my Python path. Now I'm getting the failure in ``plot_1D_marginal_probs``. Thanks for catching that. Now I've something to work with to track down this bug.

**Comment by lcgraham** [comment]
The issue is intermittent and occurs when ``save_discretization`` is called.

**Comment by smattis** [comment]
I think i see the issue. You call `comm.barrier` inside an if statement that is only called when comm.rank == 0. Every other processor will never get to this barrier.

**Comment by smattis** [comment]
That would explain the freezing.

**Comment by lcgraham** [comment]
Which if statement!?

File: bet/sample.py [hunk 3]
```diff
@@ -793,21 +843,33 @@ def save_discretization(save_disc, file_name, discretization_name=None):
...
         curr_attr = getattr(save_disc, attrname)
         if curr_attr is not None:
             new_mdat[discretization_name+attrname] = curr_attr
-    
-    if comm.rank == 0:
+   
+    if globalize and comm.rank == 0:
         if os.path.exists(file_name) or os.path.exists(file_name+'.mat'):
             mdat = sio.loadmat(file_name)
+            comm.barrier()
```
Comment (smattis) [inline bet/sample.py]: Here

**Comment by lcgraham** [comment]
I think I see the one you mean.

File: bet/sample.py [hunk 4]
```diff
@@ -793,21 +843,33 @@ def save_discretization(save_disc, file_name, discretization_name=None):
...
             for i, v in new_mdat.iteritems():
                 mdat[i] = v
             sio.savemat(file_name, mdat)
         else:
             sio.savemat(file_name, new_mdat)
+    else:
+        if os.path.exists(local_file_name) or os.path.exists(local_file_name+'.mat'):
+            mdat = sio.loadmat(local_file_name)
+            comm.barrier()
```
Comment (smattis) [inline bet/sample.py]: and here if globalize=False

**Comment by lcgraham** [comment]
Thanks Steve!

**Comment by smattis** [comment]
Still working on this? I'm getting serial and parallel tests failing.

**Comment by lcgraham** [comment]
I'm taking a break because I'm frustrated.

On Thu, Jul 7, 2016 at 5:52 PM Steven Mattis <notifications@github.com>
wrote:

> Still working on this? I'm getting serial and parallel tests failing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UT-CHG/BET/pull/243#issuecomment-231218553>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC3o7TrIecSnz1LYasU9S4axfg7Q8rHOks5qTXUYgaJpZM4JBl9Y>
> .
>
-- 

Lindley Graham, Ph.D. 2015

**Comment by smattis** [comment]
Ok. It's just that the issue in #247 is holding me back on some stuff. I don't think that part of the pull request is the issue at all. I could go fix that stuff on my own, but don't really want to introduce many conflicts. Thoughts?

**Comment by lcgraham** [comment]
How about I take 2 more days and if I can't get it fixed by then I'll
separate out the stuff that fixes #247 and move all the parallel saving
stuff to a different branch/pull request.

On Thu, Jul 7, 2016 at 6:18 PM Steven Mattis <notifications@github.com>
wrote:

> Ok. It's just that the issue in #247
> <https://github.com/UT-CHG/BET/issues/247> is holding me back on some
> stuff. I don't think that part of the pull request is the issue at all. I
> could go fix that stuff on my own, but don't really want to introduce many
> conflicts. Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/UT-CHG/BET/pull/243#issuecomment-231224957>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AC3o7bxHyeZl78SDe7B2XvU-7b0ze6qXks5qTXtQgaJpZM4JBl9Y>
> .
>
-- 

Lindley Graham, Ph.D. 2015

**Comment by lcgraham** [comment]
Does that sound ok?

On Thu, Jul 7, 2016 at 7:27 PM lcgraham <lichgraham@gmail.com> wrote:

> How about I take 2 more days and if I can't get it fixed by then I'll
> separate out the stuff that fixes #247 and move all the parallel saving
> stuff to a different branch/pull request.
>
> On Thu, Jul 7, 2016 at 6:18 PM Steven Mattis <notifications@github.com>
> wrote:
>
>> Ok. It's just that the issue in #247
>> <https://github.com/UT-CHG/BET/issues/247> is holding me back on some
>> stuff. I don't think that part of the pull request is the issue at all. I
>> could go fix that stuff on my own, but don't really want to introduce many
>> conflicts. Thoughts?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/UT-CHG/BET/pull/243#issuecomment-231224957>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe/AC3o7bxHyeZl78SDe7B2XvU-7b0ze6qXks5qTXtQgaJpZM4JBl9Y>
>> .
>>
> --
>
> Lindley Graham, Ph.D. 2015
>
-- 

Lindley Graham, Ph.D. 2015

**Comment by lcgraham** [comment]
I'm going to close this pull request for now, once I think I've got things in a better place I will either reopen it or submit a new pull request.

**Comment by smattis** [comment]
@lcgraham Try taking all of the barriers out of the tests. I don't think they are necessary, and I think they are causing the freezing.

**Comment by smattis** [comment]
I don't think nose likes them

**Comment by smattis** [comment]
Actually, from playing around, I think a barrier call anywhere can cause nose to freeze.

**Comment by lcgraham** [comment]
Nosetests is fine with the barrier calls. If you look at test_sample.py there are barrier calls which cause no problem.

**Comment by lcgraham** [comment]
I think I have things working in my fork. I'm running the full test suite in serial and parallel to confirm.

**Comment by lcgraham** [comment]
I don't know why tests are failing. My system passes both serial and parallel (multiple times). Based on where the tests are failing and the line numbers I think the code Shippable is checking is out of date. This doesn't make sense since the commit SHAs are correct.

**Comment by lcgraham** [comment]
Ready to merge! :)

**Pull request merged**